enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -6388,7 +6790,8 @@ class AdjointGenerator
       // TODO enable this if we need to free the memory
       // NOTE THAT TOPLEVEL IS THERE SIMPLY BECAUSE THAT WAS PREVIOUS ATTITUTE
       // TO FREE'ing
-      if (Mode != DerivativeMode::ReverseModeCombined) {
+      if (Mode == DerivativeMode::ReverseModeGradient ||
```

Comment:
These changes modify the behavior of reverse mode for allocations. Moreover forward mode support should be added here, which should essentially keep the same mallocs and free as the original code, duplicating as required for shadows.